Refactor the cascade

SimonSapin opened the pull request:
Converting the specified value of some properties into a computed value depends on the value of other properties. For example, the `em` unit of any length depends on the `font-size` property.

Previously, we would do a first pass over applicable declarations to build up a `values::computed::Context` struct with a number of fields for each such piece of data from other properties.

This simplies the struct by instead having it contain the set of computed values (for a given element) that is being populated and classify properties into "early" and "other", such that the only dependencies can be from "other" to "early". We iterate applicable_declarations twice, first cascading "early" properties then "other". Unfortunately, it’s not easy to check that this classification is correct.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/9929)

<!-- Reviewable:end -->

SimonSapin committed:
Fix up overflow_x and _y after the fact rather than have their values in computed::Context.

SimonSapin committed:
Fix the rem unit.

(I think it’s never worked in over a year since it was implemented…)

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/values.rs, components/style/properties.mako.rs, components/style/viewport.rs

SimonSapin commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 75382bb with merge 586f463...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/2252)

SimonSapin committed:
Refactor the cascade

Converting the specified value of some properties into a computed value
depends on the value of other properties. For example, the `em` unit
of any length depends on the `font-size` property.

Previously, we would do a first pass over applicable declarations
to build up a `values::computed::Context` struct with a number of fields
for each such piece of data from other properties.

This simplies the struct by instead having it contain the
set of computed values (for a given element) that is being populated
and classify properties into "early" and "other",
such that the only dependencies can be from "other" to "early".
We iterate applicable_declarations twice, first cascading "early" properties
then "other".
Unfortunately, it’s not easy to check that this classification is correct.

SimonSapin commented:
@bholley Alright, the last commit has the refactoring discussed today or IRC, but I’m really not confident that it’s correct **or** that our current test suite covers edge cases where is might not be correct.

Still, let’s see what happens with:

@bors-servo try

bors-servo commented:
:hourglass: Trying commit f23ed92 with merge 24115a9...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2081)

pcwalton commented on components/style/properties.mako.rs:
```diff
@@ -6798,56 +6655,92 @@ pub fn cascade(viewport_size: Size2D<Au>,
+        // Like calling to_computed_value, which wouldn't type check.
+        if style.border.border_${side}_style.none_or_hidden() &&
+           style.border.border_${side}_width != Au(0) {
+            Arc::make_mut(&mut style.border).border_${side}_width = Au(0);
```
Pushing the `Arc::make_mut()` into the if condition like you did here may improve cascading performance significantly, FWIW. Thanks!

SimonSapin replied:
Yes, that was the idea.

SimonSapin commented:
r? @pcwalton 

There’s at least one bug causing these two failures, but I don’t understand why these tests produce the results they do with this PR :(

```
  ▶ FAIL [expected PASS] /_mozilla/css/inline_absolute_hypothetical_clip_a.html
  ▶ FAIL [expected PASS] /_mozilla/css/inline_hypothetical_box_a.html

pcwalton commented:
@bors-servo: r+

Nice idea and any simplification here is helpful, since this is performance-critical code and the i-cache burden is significant.

bors-servo commented:
:pushpin: Commit f1c46c2 has been approved by `pcwalton`

<!-- @bors-servo r=pcwalton f1c46c229d116cd50d12b2016300790a84b1e26c -->

pcwalton commented:
@bors-servo: r-

Tests are failing. But r=me when they're fixed.

SimonSapin commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 2c5a19b with merge 8835648...

SimonSapin commented:
@pcwalton r? The two tests that failed in the previous try run now pass for me locally. The problem was that `-servo-display-for-hypothetical-box` needs to be "derived" from the *specified* value of `display`, not the computed value.

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2182)

SimonSapin commented:
```
  ▶ FAIL [expected PASS] /_mozilla/css/root_display_inline_a.html
  ▶ FAIL [expected PASS] /_mozilla/css/root_display_without_style_a.html
```

Looking into it now.

SimonSapin commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 6d0a07b with merge 97c13d6...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2125)

SimonSapin commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 529164e with merge 51cf187...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/2369)

SimonSapin commented:
```
  ▶ FAIL [expected PASS] /_mozilla/css/position_fixed_overflow_a.html
```

Passes locally. New intermittent?

SimonSapin commented:
Let’s see if it happens again. I’m also trying chaos mode locally.

@bors-servo try

SimonSapin commented:
@bors-servo retry
?

bors-servo commented:
:hourglass: Trying commit 529164e with merge 28557d8...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/2055), [gonk](http://build.servo.org/builders/gonk/builds/2111), [linux-dev](http://build.servo.org/builders/linux-dev/builds/2089), [linux-rel](http://build.servo.org/builders/linux-rel/builds/2370), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/782), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2138), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2196), [status-appveyor](https://ci.appveyor.com/project/servo/servo/build/1.0.662)

jdm commented:
@SimonSapin: #9012

SimonSapin commented:
Ok, I’ll remember to try more variations of github search.

@pcwalton try passed, please r? the bug fixes, I’m not super confident in them.

pcwalton commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[components/style/properties.mako.rs, line 6681 \[r5\]](https://reviewable.io:443/reviews/servo/servo/9929#-KDAxCfYzRg25FIiaahZ:-KDAxCfYzRg25FIiaah_:432850299) ([raw file](https://github.com/servo/servo/blob/529164e4a5b0fcca67d5a1fa31f17f87c038eadd/components/style/properties.mako.rs#L6681)):*
Does it make sense to only call `Arc::make_mut()` if we're actually going to make a change? `Arc::make_mut()` is always high in the profile.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9929)*
<!-- Sent from Reviewable.io -->

SimonSapin commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[components/style/properties.mako.rs, line 6681 \[r5\]](https://reviewable.io:443/reviews/servo/servo/9929#-KDAxCfYzRg25FIiaahZ:-KDAxgVsAMDaEmttV3qt:-1338306070) ([raw file](https://github.com/servo/servo/blob/529164e4a5b0fcca67d5a1fa31f17f87c038eadd/components/style/properties.mako.rs#L6681)):*
Yes, we only go inside `if let Some(…` if we’re going to make a change. The `match` above is `None` for no change.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9929)*
<!-- Sent from Reviewable.io -->

pcwalton commented:
Looks good, r=me.

SimonSapin commented:
@bors-servo r=pcwalton

bors-servo commented:
:pushpin: Commit 529164e has been approved by `pcwalton`

<!-- @bors-servo r=pcwalton 529164e4a5b0fcca67d5a1fa31f17f87c038eadd -->

bors-servo commented:
:hourglass: Testing commit 529164e with merge 1ade640...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/2376)

bholley commented:
@bors-servo retry

mbrubeck commented:
The failed test was #9724.

bors-servo commented:
:zap: Previous build results for [android](http://build.servo.org/builders/android/builds/2060), [gonk](http://build.servo.org/builders/gonk/builds/2117), [linux-dev](http://build.servo.org/builders/linux-dev/builds/2095), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/787), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2144), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2202), [status-appveyor](https://ci.appveyor.com/project/servo/servo/build/1.0.681) are reusable. Rebuilding only [linux-rel](http://build.servo.org/builders/linux-rel/builds/2376)...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/2378)

KiChjang commented:
@bors-servo retry
* #9818

bors-servo commented:
:hourglass: Testing commit 529164e with merge 2d6283c...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/2063), [gonk](http://build.servo.org/builders/gonk/builds/2120), [linux-dev](http://build.servo.org/builders/linux-dev/builds/2097), [linux-rel](http://build.servo.org/builders/linux-rel/builds/2379), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/789), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2146), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2204), [status-appveyor](https://ci.appveyor.com/project/servo/servo/build/1.0.691)

bholley commented on components/style/properties.mako.rs:
```diff
@@ -6498,23 +6457,17 @@ fn cascade_with_cached_declarations(
                                 % endif
 
                                 % if property.name in DERIVED_LONGHANDS:
-                                    % if not style_struct.inherited:
```
@SimonSapin what happened to this bit?

SimonSapin replied:
It’s not needed since `derive_from_*` functions don’t take a `computed_value` parameter anymore. Computed values (as known "so far") are already available through e.g. `context.style.box_.display`, whereas `context.display` was previously the *specified* value, so when the computed value was needed it had to be passed separately.

bholley replied:
I was talking about the |if not style_struct.inherited| bit.

SimonSapin replied:
Yes. In this function non-inherited styles are cached so we don’t need to run the cascade for them again. `let computed_value = …;` is in a larger block above is in `% if style_struct.inherited:` (*without* `not`).

Before this PR, `computed_value` was needed unconditionally, so this `% if not style_struct.inherited:` block (*with* `not`) defined it for the rest of the cases. Now that `computed_value` is not used in the `derive_from_*` call anymore, this second `let computed_value = …;` statement would be dead code.
